sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java

```diff
@@ -0,0 +1,569 @@
+        name = null;
+      }
+      if (fullPath) {
+        fullPath = false;
+        return new S3Path(innerFileSystem, component, "");
+      } else {
+        // Relative paths have no bucket.
+        return new S3Path(innerFileSystem, "", component);
+      }
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  public S3Path commonKeyPrefix(S3Path other) {
+    checkArgument(getBucket().equals(other.getBucket()), "buckets equal");
+
+    int minLength = Math.min(getKey().length(), other.getKey().length());
+    String commonKeyPrefix = getKey().substring(0, minLength);
+    for (int i = 0; i < minLength; i++) {
+      if (getKey().charAt(i) != other.getKey().charAt(i)) {
+        commonKeyPrefix = getKey().substring(0, i);
+        break;
+      }
+    }
+
+    return new S3Path(null, getBucket(), commonKeyPrefix);
+  }
+
+  @Override
+  public int compareTo(Path other) {
+    if (!(other instanceof S3Path)) {
+      throw new ClassCastException();
+    }
+
+    S3Path path = (S3Path) other;
+    int b = bucket.compareTo(path.bucket);
+    if (b != 0) {
+      return b;
+    }
+
+    // Compare a component at a time, so that the separator char doesn't
+    // get compared against component contents.  Eg, "a/b" < "a-1/b".
+    Iterator<Path> left = iterator();
+    Iterator<Path> right = path.iterator();
+
+    while (left.hasNext() && right.hasNext()) {
+      String leftStr = left.next().toString();
+      String rightStr = right.next().toString();
+      int c = leftStr.compareTo(rightStr);
+      if (c != 0) {
+        return c;
+      }
+    }
+
+    if (!left.hasNext() && !right.hasNext()) {
+      return 0;
+    } else {
+      return left.hasNext() ? 1 : -1;
+    }
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    S3Path otherS3Path = (S3Path) other;
+    return bucket.equals(otherS3Path.bucket) && key.equals(otherS3Path.key);
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * bucket.hashCode() + key.hashCode();
+  }
+
+  @Override
+  public String toString() {
+    if (!isAbsolute()) {
+      return key;
+    }
+    StringBuilder sb = new StringBuilder();
+    sb.append(SCHEME).append("://");
+    if (!bucket.isEmpty()) {
+      sb.append(bucket).append('/');
+    }
+    sb.append(key);
+    return sb.toString();
+  }
+
+  private String bucketAndKey() {
+    if (bucket.isEmpty()) {
+      return key;
+    }
+    return bucket + "/" + key;
+  }
+
+  public String getKeyNonWildcardPrefix() {
+    Matcher m = GLOB_PREFIX.matcher(getKey());
+    checkArgument(m.matches(), String.format("Glob expression: [%s] is not expandable.", getKey()));
+    return m.group("PREFIX");
+  }
+
+  /**
+   * A coder for {@link S3Path}.
+   */
+  public static class Coder extends SerializableCoder<S3Path> {
```

Comment:
I'd strongly recommend against extending a concrete class. S3Path already implements Serializable, so this class should just be unnecessary.

Reply:
Consolidated S3Path into S3ResourceId, and eliminated the coder. Users specify paths as strings, so this shouldn't have been necessary.